### ISSUE TITLE
ProtoLibrarySupport: Basic support for indexing proto field usage in parsed string literals. [phab:D1274]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1656]
_Created by **courbet** at 2017-01-24 13:04:18_: 

See kythe/cxx/indexer/cxx/testdata/libraries/proto_parsetextproto.cc for
an example.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

Nit: 2017.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

This isn't actually the case. We actually get much more accurate information from protoc to link generated proto code directly back to the source .proto files. This will still work with the mechanism you're using, since the accessors you're linking to will transitively link to the relevant part of the message definitions.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

ultranit: please line up that (

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

Is there a reason you're doing a linear walk here instead of using `DeclContext::lookup()`?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

See the comment below about DeclContext::lookup().

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

There exists `ASTContext::getTranslationUnitDecl()`.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

It would be better to use the `NodeId` of the function that contains the string literal, not an ID for the literal itself. This corresponds with your treatment of the literal as syntactic sugar for a bunch of protobuf function calls. (It also allows you to use the existing logic for generating signatures for functions so that these call sites won't appear weirdly in a call graph.)

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

Why is this boxed?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

Could you make this into a static member function instead? That way you wouldn't have to worry about a ParseTextProtoHandler being put into an unusable state.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-26 22:24:48_:

You might want to guard against non-UTF-8 strings here. `\r\n` is a popular encoding for endline on Windows; some people use `\r`.

**Comment by kythe** [comment]
_**courbet** wrote at 2017-01-27 14:07:53_:

  - Address review comments

**Comment by kythe** [comment]
_**courbet** commented inline (context lost during migration) at 2017-01-27 14:08:21_:

Nope, thanks for the suggestion.

**Comment by kythe** [comment]
_**courbet** commented inline (context lost during migration) at 2017-01-27 14:08:21_:

That was the point I was trying to explain. I've rephrased the comment, is that better ?

**Comment by kythe** [comment]
_**courbet** commented inline (context lost during migration) at 2017-01-27 14:08:21_:

Actually the io::Tokenizer increments lines on '\n' only:

https://github.com/google/protobuf/blob/master/src/google/protobuf/io/tokenizer.cc#L226

**Comment by kythe** [comment]
_**courbet** commented inline (context lost during migration) at 2017-01-27 14:11:54_:

Can you elaborate a bit on that ? I'm not very familiar with how kythe generates Ids.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-27 21:46:08_:

The comment is clearer (thanks!), but the problem bit is "the proto indexer outputs searchable names". (It could just be that "searchable names" is a term of art for me and not for you.) The proto indexer and the proto compiler collaborate through metadata to link generated code back to the protobuf definitions.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-27 21:46:08_:

Uff da. You've reminded me of protoc's "tabs = up to 8 columns, depending" rule too.

Please add a comment about this (e.g. "We're building this table so that we can map io::Tokenizer lines and columns back to byte offsets in the string literal.")

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-27 21:46:08_:

Will this do something surprising if the line contains tabs?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-27 21:46:08_:

Sure. The indexer can generate an ID for every Decl (::BuildNodeIdForDecl), any range of source text ("anchors"), and for most other parts of the AST. These eventually identify nodes in the Kythe graph. When we want to represent a callsite, there are three nodes involved: the one for the range of text covering the call, the one for the function being called, and the one for the function that should be blamed for doing the calling. The indexer maintains a stack of these nodes. Right now your code is blaming the call on the ASCII proto literal, which is probably not the right thing to do. Instead, I suggest using the ID on the top of the blame stack; see the code for [[ https://github.com/google/kythe/blob/master/kythe/cxx/indexer/cxx/IndexerASTHooks.cc#L1280 | VisitCallExpr ]].

**Comment by kythe** [comment]
_**courbet** wrote at 2017-01-30 10:26:25_:

  - More comments.
  - Blame field access with the regular stack system.
  - Add a unit test with tabs and fix it.

**Comment by kythe** [comment]
_**courbet** commented inline (context lost during migration) at 2017-01-30 10:28:18_:

Thanks, I've updated the comment.

**Comment by kythe** [comment]
_**courbet** commented inline (context lost during migration) at 2017-01-30 10:28:18_:

Thanks for the explanation. Makes sense.

**Comment by kythe** [comment]
_**courbet** commented inline (context lost during migration) at 2017-01-30 10:28:18_:

It does, thanks for the catch. I've added unit test and I'm now handling tabs (though the approach is not great). I'm going to do a PR against google/protobuf to handle that "correctly".

**Comment by kythe** [comment]
_**zarko** wrote at 2017-02-03 04:37:49_:

It doesn't look like there are any rules to compile ProtoLibrarySupport.cc now (and the code that registers the plugin was removed). Did something go wrong generating the diff?

I'm unable to apply the patch to my checkout to run the tests--I think you need to rebase.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-02-06 18:56:21_:

For reference, this is the warning I get when I try to apply the patch:
```
    This diff is against commit 923c9579fbff7a396912699e7581de2c78615f6c, but
    the commit is nowhere in the working copy. Try to apply it against the
    current working copy state? (fd5f94f2ab6f92dc70dc736bfda662cc33131d8f)
    [Y/n] 
```
and then the patch fails. The previous comments about the build rules/plugin registration going missing still apply.

**Comment by kythe** [comment]
_**courbet** wrote at 2017-02-07 15:08:40_:

  - Address review comments
  - More comments.
  - Blame field access with the regular tack system.
  - Add a unit test with tabs and fix it.

**Comment by kythe** [comment]
_**courbet** wrote at 2017-02-07 16:12:44_:

Try again.

**Comment by kythe** [comment]
_**courbet** wrote at 2017-02-07 16:17:10_:

OK, sorry for the noise. I guess I don't understand how arcanist works. I got it to work by explicitly specifying against which commit to diff (arc diff 088b53f5). I expected it to track the rebase after the first export, but that doe snot seem to be happening.

**Comment by kythe** [comment]
_**courbet** wrote at 2017-02-07 16:30:56_:

Fix bad merge.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-02-08 19:02:20_:

I'm not entirely sure what you did, but now the patch is pulling in unrelated files; did `arc diff --update 1274` do something unexpected? You'll still need to rebase your git branch, which is separate from what Arcanist is tracking.

**Comment by kythe** [comment]
_**zarko** required changes at 2017-02-10 18:53:18_

**Comment by kythe** [comment]
_**courbet** wrote at 2017-02-15 16:24:24_:

  - Address review comments
  - More comments.
  - Blame field access with the regular stack system.
  - Add a unit test with tabs and fix it.

**Comment by kythe** [comment]
_**courbet** wrote at 2017-02-15 16:27:56_:

Again

**Comment by kythe** [comment]
_**courbet** wrote at 2017-02-15 16:31:12_:

Giving up arcanist and using a raw diff.

**Comment by kythe** [comment]
_**courbet** wrote at 2017-02-15 16:32:27_:

I have no idea what happened with arcanist. I've uploaded a raw diff (git diff -U999999 remote/master), please tell me if that works.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-02-15 18:47:28_:

OK, I got that patched in. Thanks for dealing with `arc`.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-02-15 18:54:55_

**Comment by kythe** [comment]
_**zarko** closed via commit 750d8e82bf3f33f1d21647c8dba0830df31ffdf2 at 2017-02-15 18:55:31_
